src/NServiceBus.Core/Transports/Development/DevelopmentTransportMessagePump.cs

```diff
@@ -0,0 +1,306 @@
+            }
+
+            concurrencyLimiter.Dispose();
+            runningReceiveTasks.Clear();
+        }
+
+        [DebuggerNonUserCode]
+        async Task ProcessMessages()
+        {
+            try
+            {
+                await InnerProcessMessages().ConfigureAwait(false);
+            }
+            catch (OperationCanceledException)
+            {
+                // For graceful shutdown purposes
+            }
+            catch (Exception ex)
+            {
+                Logger.Error("File Message pump failed", ex);
+                //await peekCircuitBreaker.Failure(ex).ConfigureAwait(false);
+            }
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                await ProcessMessages().ConfigureAwait(false);
+            }
+        }
+
+        async Task InnerProcessMessages()
+        {
+            while (!cancellationTokenSource.IsCancellationRequested)
+            {
+                var filesFound = false;
+
+                foreach (var filePath in Directory.EnumerateFiles(path, "*.*"))
+                {
+                    filesFound = true;
+
+                    var nativeMessageId = Path.GetFileNameWithoutExtension(filePath);
+
+                    IDevelopmentTransportTransaction transaction;
+
+                    if (transactionMode != TransportTransactionMode.None)
+                    {
+                        transaction = new DirectoryBasedTransaction(path);
+                    }
+                    else
+                    {
+                        transaction = new NoTransaction();
+                    }
+
+                    transaction.BeginTransaction(filePath);
+
+                    await concurrencyLimiter.WaitAsync(cancellationToken).ConfigureAwait(false);
+
+                    var task = Task.Run(async () =>
+                    {
+                        try
+                        {
+                            await ProcessFile(transaction, nativeMessageId).ConfigureAwait(false);
+
+                            transaction.Complete();
+
+                            receiveCircuitBreaker.Success();
+                        }
+                        catch (Exception ex)
+                        {
+                            await receiveCircuitBreaker.Failure(ex).ConfigureAwait(false);
+                        }
+                        finally
+                        {
+                            concurrencyLimiter.Release();
+                        }
+                    }, cancellationToken);
+
+                    task.ContinueWith(t =>
+                        {
+                            Task toBeRemoved;
+                            runningReceiveTasks.TryRemove(t, out toBeRemoved);
+                        }, TaskContinuationOptions.ExecuteSynchronously)
+                        .Ignore();
+
+                    runningReceiveTasks.AddOrUpdate(task, task, (k, v) => task)
```

Comment:
You could use `TryAdd `then you don't need the `Ignore`. Depends on the outcome of the whether we need the tracking discussion above